internal/access/identity.go

```diff
@@ -86,19 +86,25 @@ func ListIdentities(c *gin.Context, opts data.ListIdentityOptions) ([]models.Ide
 	return data.ListIdentities(db, opts)
 }
 
-func GetContextProviderIdentity(c RequestContext) (*models.Provider, string, error) {
+func GetContextProviderIdentity(c RequestContext, google *models.Provider) (*models.Provider, string, error) {
 	// does not need authorization check, this action is limited to the calling user
-	provider, err := data.GetProvider(c.DBTxn, data.GetProviderOptions{
-		ByID: c.Authenticated.AccessKey.ProviderID,
-	})
-	if err != nil {
-		return nil, "", fmt.Errorf("user info provider: %w", err)
-	}
+	var provider *models.Provider
+	if c.Authenticated.AccessKey.ProviderID == google.ID {
+		provider = google
```

Comment:
skip looking this up in the database

Reply:
This `GetContextProviderIdentity` function only has 1 caller, right?

The logic is a bit awkward right now because we pass in a provider , but only use it in some cases and we could have figured out that case from the caller.

If we moved these 2 lines and the `GetProvider` logic below to the caller I think it would be easier to follow:

```go
var provider *models.Provider
if rCtx.Authenticated.AccessKey.ProviderID == models.InternalGoogleProviderID {
    provider = a.Server.google
} else {
    var err error
    provider, err = data.GetProvider(...)
    ...
}

// This branch is in two places right now, it could be in only one of those places if we move the logic to the caller
if provider.Kind == models.ProviderKindInfra {
    ...
}

providerUser, err := data.GetProviderUser(c.DBTxn, c.Authenticated.AccessKey.ProviderID, identity.ID)
if err != nil {
    return nil, "", err
}
```

This would allow us to remove `access.GetContextProviderIdentity` entirely.

I know why we didn't do that originally, because we have this rule about "`data` functions can only be called from `access`". This is a great example of how that rule is causing is to write code that is harder to read and maintain.

The code you have here looks correct to me, so I'm ok if we want to explore that change in a separate PR.